conda/core/index.py

```diff
@@ -331,7 +333,61 @@ def maybe_decompress(filename, resp_content):
                              getattr(e.response, 'elapsed', None))
 
 
-def read_local_repodata(cache_path):
+def write_pickled_repodata(cache_path, repodata):
+    # Don't bother to pickle empty channels
+    if not repodata.get('packages'):
+        return
+    try:
+        with open(cache_path + '.q', 'wb') as f:
+            pickle.dump(repodata, f)
+    except:
+        pass
+
+
+def read_picked_repodata(cache_path, channel_url, schannel, priority):
+    pickle_path = cache_path + '.q'
+    # Don't trust pickled data if there is no accompanying json data
+    if not isfile(pickle_path) or not isfile(cache_path):
+        return None
+    try:
+        with open(cache_path + '.q', 'rb') as f:
+            repodata = pickle.load(f)
+    except:
+        return None
+    if repodata.get('_url') != channel_url:
+        return None
+    add_pip = context.add_pip_as_python_dependency
+    changed = False
+    # Safe changes to IndexRecord
+    if (repodata['_add_pip'] != add_pip or 
+        repodata['_priority'] != priority or
+        repodata['_schannel'] != schannel):
+        repodata['_priority'] = priority
+        repodata['_add_pip'] = add_pip
+        changed = True
+        for rec in itervalues(repodata.get('packages', {})):
+            rec.priority = priority
+            rec.schannel = schannel
+            if rec.name == 'python' and rec.version.startswith(('2.', '3.')):
+                if add_pip:
+                    rec.depends = rec.depends + ('pip',)
+                elif rec.depends[-1] == 'pip':
+                    rec.depends = rec.depends[:-1]
+    # Safe changes to Dist
+    if repodata['_schannel'] != schannel:
```

Comment:
Recompute the Dist keys if necessary. This should only happen if `channel_alias` is modified, so that a "short" channel becomes a "long" channel or vice versa.